core/src/main/java/io/grpc/inprocess/InProcessTransport.java

```diff
@@ -116,14 +116,17 @@ public void run() {
 
   @Override
   public synchronized ClientStream newStream(MethodDescriptor<?, ?> method,
-      Metadata headers, ClientStreamListener clientStreamListener) {
+      Metadata headers) {
     if (shutdownStatus != null) {
-      clientStreamListener.closed(shutdownStatus, new Metadata());
-      return new NoopClientStream();
+      return new NoopClientStream() {
+        @Override
+        public void start(ClientStreamListener listener) {
+          listener.closed(shutdownStatus, new Metadata());
+        }
+      };
     }
     headers.removeAll(GrpcUtil.AUTHORITY_KEY);
     InProcessStream stream = new InProcessStream();
-    stream.serverStream.setListener(clientStreamListener);
```

Comment:
If this is delayed, then `serverTransportListener.streamCreated` needs to be delayed, and so `clientStream.setListener` needs to be delayed, and that causes a race with `shutdownStatus`, such that it should be delayed. So basically everything should be delayed (it doesn't matter when the `headers.removeAll` is done, but might as well delay it with everything else).

I think that leaves in this method just `new InProcessStream()` and some lines to save `method` and `headers`.